Source/Core/VideoBackends/D3D/DXShader.cpp

```diff
@@ -0,0 +1,197 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <algorithm>
+#include <cstddef>
+
+#include "Common/Assert.h"
+#include "Common/CommonTypes.h"
+#include "Common/Logging/Log.h"
+
+#include "VideoBackends/D3D/D3DBase.h"
+#include "VideoBackends/D3D/D3DShader.h"
+#include "VideoBackends/D3D/D3DState.h"
+#include "VideoBackends/D3D/D3DUtil.h"
+#include "VideoBackends/D3D/DXShader.h"
+#include "VideoBackends/D3D/Render.h"
+
+#include "VideoCommon/ImageWrite.h"
+#include "VideoCommon/TextureConfig.h"
+
+namespace DX11
+{
+DXShader::DXShader(D3DBlob* bytecode, ID3D11VertexShader* vs)
+    : AbstractShader(ShaderStage::Vertex), m_bytecode(bytecode), m_shader(vs)
+{
+}
+
+DXShader::DXShader(D3DBlob* bytecode, ID3D11GeometryShader* gs)
+    : AbstractShader(ShaderStage::Geometry), m_bytecode(bytecode), m_shader(gs)
+{
+}
+
+DXShader::DXShader(D3DBlob* bytecode, ID3D11PixelShader* ps)
+    : AbstractShader(ShaderStage::Pixel), m_bytecode(bytecode), m_shader(ps)
+{
+}
+
+DXShader::DXShader(D3DBlob* bytecode, ID3D11ComputeShader* cs)
+    : AbstractShader(ShaderStage::Compute), m_bytecode(bytecode), m_shader(cs)
+{
+}
+
+DXShader::~DXShader()
+{
+  m_shader->Release();
+  m_bytecode->Release();
+}
+
+ID3D11VertexShader* DXShader::GetD3DVertexShader() const
+{
+  _assert_(m_stage == ShaderStage::Vertex);
+  return static_cast<ID3D11VertexShader*>(m_shader);
+}
+
+ID3D11GeometryShader* DXShader::GetD3DGeometryShader() const
+{
+  _assert_(m_stage == ShaderStage::Geometry);
+  return static_cast<ID3D11GeometryShader*>(m_shader);
+}
+
+ID3D11PixelShader* DXShader::GetD3DPixelShader() const
+{
+  _assert_(m_stage == ShaderStage::Pixel);
+  return static_cast<ID3D11PixelShader*>(m_shader);
+}
+
+ID3D11ComputeShader* DXShader::GetD3DComputeShader() const
+{
+  _assert_(m_stage == ShaderStage::Compute);
+  return static_cast<ID3D11ComputeShader*>(m_shader);
+}
+
+bool DXShader::HasBinary() const
+{
+  // m_bytecode should always be non-null.
+  return true;
+}
+
+AbstractShader::BinaryData DXShader::GetBinary() const
+{
+  BinaryData ret;
+  ret.resize(m_bytecode->Size());
+  std::memcpy(ret.data(), m_bytecode->Data(), ret.size());
+  return ret;
+}
+
+std::unique_ptr<AbstractShader> Renderer::CreateShaderFromSource(ShaderStage stage,
+                                                                 const char* source, size_t length)
+{
+  D3DBlob* bytecode;
```

Comment:
Maybe make this function leverage `CreateProgramFromBinary` or have them both use the same functionality somehow?